### PR TITLE
Surface force

### PR DESCRIPTION
**Pull request by Rozie100**
Surface force calculation is added.

### CODE REVIEW

**Review by mmcgurn — COMMENTED** [review]
Thanks for putting this together.  I'm a little confused by a few things, but most of it is why there is separate 2D/3D sections.  What I would recommend doing is to write out the math/algorithm we are trying to implement (this can be in the document or in the method comment).  I'm having a hard time understanding why we have dim specific code, I think we may want to break this is to a setup call that builds the stencil and then a separate evaluate call.  I did something similar when using (https://github.com/mmcgurn/ablate/blob/b848aabf9b1b0f141bd47dca68c27170020b4a94/src/finiteVolume/stencils/stencil.hpp
https://github.com/UBCHREST/ablate/blob/b848aabf9b1b0f141bd47dca68c27170020b4a94/src/finiteVolume/stencils/stencil.hpp
https://github.com/danial-faghihi/ablate/blob/b848aabf9b1b0f141bd47dca68c27170020b4a94/src/finiteVolume/stencils/stencil.hpp
https://github.com/owenmylotte/ablate/blob/b848aabf9b1b0f141bd47dca68c27170020b4a94/src/finiteVolume/stencils/stencil.hpp
https://github.com/ubdsgroup/ablate/blob/b848aabf9b1b0f141bd47dca68c27170020b4a94/src/finiteVolume/stencils/stencil.hpp
https://github.com/Rozie100/ablate/blob/b848aabf9b1b0f141bd47dca68c27170020b4a94/src/finiteVolume/stencils/stencil.hpp
https://github.com/jpdening/ablate/blob/b848aabf9b1b0f141bd47dca68c27170020b4a94/src/finiteVolume/stencils/stencil.hpp
https://github.com/maesemen/ablate/blob/b848aabf9b1b0f141bd47dca68c27170020b4a94/src/finiteVolume/stencils/stencil.hpp
https://github.com/klbudzin/ablate/blob/b848aabf9b1b0f141bd47dca68c27170020b4a94/src/finiteVolume/stencils/stencil.hpp)

File: src/finiteVolume/processes/surfaceForce.cpp [hunk 1]
```diff
@@ -0,0 +1,397 @@
...
+                                        // get centroid for attached-cells to each face
+
+                                        PetscFVCellGeom* cg;
+                                        DMPlexPointLocalRead(dmCell, cell[iCell], cellGeomArray, &cg);
+                                        // get current alpha of each attached face
+                                        PetscScalar* alpha = nullptr;
+                                        DMPlexPointLocalFieldRead(dm, cell[iCell], VFfield->id, solArray, &alpha);
+
+                                        for (PetscInt d = 0; d < dim; ++d) {
+                                            if (cg->centroid[d] > xyz[d]) {
```
[the remark this answers is not in the document]
Comment (Rozie100) [reply src/finiteVolume/processes/surfaceForce.cpp:309]: If the centroid falls forward or behind the reference vertex, it changes the sign of contribution for the alpha gradient. The distance of the two centers is added up by summing their distance to the reference vertex.

File: src/finiteVolume/processes/surfaceForce.cpp [hunk 2]
```diff
@@ -0,0 +1,397 @@
...
+                                                verNormal[d] = *alpha;
+
+                                            } else if (cg->centroid[d] < xyz[d]) {
+                                                verNormal[d] = -*alpha;
+
+                                            } else {
+                                                verNormal[d] = 0;
+                                            }
+                                            distance[d] += abs(cg->centroid[d] - xyz[d] + TINY);
```
[the remark this answers is not in the document]
Comment (Rozie100) [reply src/finiteVolume/processes/surfaceForce.cpp:318]: If aligned, there is no contribution to the gradient at all. If a small distance, it should be moderated by the rest of calculations.

File: src/finiteVolume/processes/surfaceForce.cpp [hunk 3]
```diff
@@ -0,0 +1,397 @@
...
+
+                                DMPlexGetSupport(dm, edges[iEdges], &attachedFace);
+                                DMPlexGetSupportSize(dm, edges[iEdges], &numAttachedFace);
+
+                                // Get cells attached to each face
+                                for (PetscInt iFace = 0; iFace < numAttachedFace; ++iFace) {
+                                    const PetscInt* cell;
+                                    PetscInt numCell;
+
+                                    DMPlexGetSupport(dm, attachedFace[iFace], &cell);
```
[the remark this answers is not in the document]
Comment (Rozie100) [reply src/finiteVolume/processes/surfaceForce.cpp:290]: That's correct. Based on the geometry, each cell with be used " dim" times; therefore final normal is divided by "dim".

File: src/finiteVolume/processes/surfaceForce.cpp [hunk 4]
```diff
@@ -0,0 +1,397 @@
...
+        const PetscInt* cellFaces;
+        PetscInt numCellFaces;
+        DMPlexGetCone(dm, c, &cellFaces);
+        DMPlexGetConeSize(dm, c, &numCellFaces);
+        // extract number of vortices of the cell for 2D
+        PetscReal numVertex;
+        numVertex = numCellFaces;
+        // calculate number of vortices of the cell for 3D
+        PetscReal vertex = 0;
+        PetscReal numVortices;
```
[the remark this answers is not in the document]
Comment (Rozie100) [reply src/finiteVolume/processes/surfaceForce.cpp:93]: Both use "numVertex". Just more steps/ variable takes for 3D to get the final "numVertex".

File: src/finiteVolume/processes/surfaceForce.cpp [hunk 5]
```diff
@@ -0,0 +1,397 @@
...
+                // find edges attached to faces of the cell
+                DMPlexGetCone(dm, cellFaces[face], &attachEdges);
+                DMPlexGetConeSize(dm, cellFaces[face], &numAttachEdges);
+                vertex += numAttachEdges;
+                numVortices = vertex / dim;
+            }
+        }
+        // get vortices attached to each face
+        for (PetscInt f = 0; f < numCellFaces; f++) {
+            // here if the case is 2D extract vortices, if 3D extract edges then proceed
```
[the remark this answers is not in the document]
Comment (Rozie100) [reply src/finiteVolume/processes/surfaceForce.cpp:108]: Here there is one more level to extract vortices for 3D.

File: src/finiteVolume/processes/surfaceForce.cpp [hunk 6]
```diff
@@ -0,0 +1,303 @@
...
+                DMPlexPointLocalRead(cdm, vertex, normalArray, &vertexNormal);
+                xyz[d] = coordsArray[offset + d];
+
+                // calculate normal at each cell center using normals of connected vortices to the cell
+                centerNormal[d] += vertexNormal[d];
+
+                // calculate divergence of normal for each cell center using vertex normals --> Delta.ni,j=  [nx(i+1/2,j+1/2)-nx(i-1/2,j+1/2)+nx(i+1/2,j-1/2)-nx(i-1/2,j-1/2)]/2*deltaX +
+                // [ny(i+1/2,j+1/2)-ny(i+1/2,j-1/2)+ny(i-1/2,j+1/2)-ny(i-1/2,j-1/2)]/2*deltaY
+                // get magnitude of normals at vortices to calculate the derivative of the magnitude of normals at the cell center
+                distance[d] += abs(xyz[d] - fcg->centroid[d] + utilities::Constants::tiny);
```
Comment (mmcgurn) [inline src/finiteVolume/processes/surfaceForce.cpp:249]: Couldn't this distance calculation be done in the stencil so you would not need to get xyz each time?

**Review by mmcgurn — CHANGES_REQUESTED** [review]

File: src/finiteVolume/processes/surfaceForce.cpp (hunk 6 above)
Comment (Rozie100) [reply src/finiteVolume/processes/surfaceForce.cpp:249]: I am only marching over cells here, not in the setup, so probably just do it here.
Regarding close cells line up, I checked the code again, and I calculated weights everywhere that gradients are tasked as the original ALE- scheme suggests. So I believe there may not be any fake gradients or singularity problems.
....
Still waiting to hear back from Dr. Keneply!

File: src/finiteVolume/processes/CMakeLists.txt [hunk 7]
```diff
@@ -13,7 +13,7 @@ target_sources(ablateLibrary
         les.cpp
         chemistry.cpp
         radiationLoss.cpp
-        thermophoreticDiffusion.cpp
+        surfaceForce.cpp
```
[the remark this answers is not in the document]
Comment (Rozie100) [reply src/finiteVolume/processes/CMakeLists.txt:17]: Sorry, not intentional!

**Review by mmcgurn — CHANGES_REQUESTED** [review]
Looks good,  I think we would want to move the dmClone and field setup to the setup function so it only needs to be done once.

File: src/finiteVolume/processes/surfaceForce.cpp [hunk 8]
```diff
@@ -0,0 +1,297 @@
...
+
+    // get the coordinate domain
+    DM cdm;
+    DMGetCoordinateDM(dm, &cdm);
+
+    // create a domain to get a local vector to save normals
+    DM dmData;
+    PetscFE fe_coords;
+    PetscInt k = 1;
+    DMClone(dm, &dmData);
```
[the remark this answers is not in the document]
Comment (Rozie100) [reply src/finiteVolume/processes/surfaceForce.cpp]: Done! the test looks much better now. Thanks!

**Review by mmcgurn — APPROVED** [review]

**Pull request merged**